server/plugins/LFUGens.cpp

```diff
@@ -3228,56 +3231,24 @@ void IEnvGen_Ctor(IEnvGen* unit) {
         SETCALC(IEnvGen_next_k);
     }
 
-    // pointer, offset
-    // initlevel, numstages, totaldur,
-    // [dur, shape, curve, level] * numvals
     int numStages = (int)IN0(3);
     int numvals = numStages * 4; // initlevel + (levels, dur, shape, curves) * stages
-    float offset = unit->m_offset = IN0(1);
-    float point = unit->m_pointin = IN0(0) - offset;
     unit->m_envvals = (float*)RTAlloc(unit->mWorld, (int)(numvals + 1.) * sizeof(float));
     ClearUnitIfMemFailed(unit->m_envvals);
 
+    // fill m_envvals with the values
     unit->m_envvals[0] = IN0(2);
-    //  Print("offset of and initial  values %3,3f, %3.3f\n", offset, unit->m_envvals[0]);
-    // fill m_envvals with the values;
     for (int i = 1; i <= numvals; i++) {
         unit->m_envvals[i] = IN0(4 + i);
-        //      Print("val for: %d, %3.3f\n", i, unit->m_envvals[i]);
     }
 
-    //  float out = OUT0(0);
-    float totalDur = IN0(4);
-    float level = 0.f;
-    float newtime = 0.f;
-    int stage = 0;
-    float seglen = 0.f;
-    if (point >= totalDur) {
-        unit->m_level = level = unit->m_envvals[numStages * 4]; // grab the last value
-    } else {
-        if (point <= 0.0) {
-            unit->m_level = level = unit->m_envvals[0];
-        } else {
-            float segpos = point;
-            // determine which segment the current time pointer needs calculated
-            for (int j = 0; point >= newtime; j++) {
-                seglen = unit->m_envvals[(j * 4) + 1];
-                newtime += seglen;
-                segpos -= seglen;
-                stage = j;
-            }
+    unit->m_offset = IN0(1);
+    unit->m_pointin = uninitializedControl; // trigger calculation in next_k
+    unit->m_level = uninitializedControl; // will be set in next_k
 
-            segpos = segpos + seglen;
-            float begLevel = unit->m_envvals[(stage * 4)];
-            int shape = (int)unit->m_envvals[(stage * 4) + 2];
-            int curve = (int)unit->m_envvals[(stage * 4) + 3];
-            float endLevel = unit->m_envvals[(stage * 4) + 4];
-            float pos = (segpos / seglen);
+    IEnvGen_next_k(unit, 1);
 
-            GET_ENV_VAL
-        }
-    }
-    OUT0(0) = level;
+    unit->m_pointin = uninitializedControl; // trigger calculation in next_x
```

Comment:
Then this state variable is set to `uninitializedControl` so that once the `next` function is called, the state will be re-calculated.